File: backend/app/schemas.py
from typing import List, Literal, Optional

from pydantic import BaseModel

PAYMENT_METHODS = Literal["Tunai", "BCA TOKO", "BRI TOKO", "BCA ADMIN (ELIS)", "QRIS", "E-Wallet"]
ROLES = Literal["Owner", "Manager", "Kasir", "Gudang"]


class LoginInput(BaseModel):
    username: str
    password: str


class ChangePasswordInput(BaseModel):
    current_password: str
    new_password: str


class UserCreate(BaseModel):
    username: str
    password: str
    name: str
    role: ROLES


class UserUpdate(BaseModel):
    name: Optional[str] = None
    role: Optional[ROLES] = None
    password: Optional[str] = None
    active: Optional[bool] = None


class CategoryInput(BaseModel):
    name: str
    color: Optional[str] = "#2563EB"
    image: Optional[str] = ""


class ProductInput(BaseModel):
    name: str
    sku: Optional[str] = ""
    barcode: Optional[str] = ""
    category_id: Optional[str] = None
    price: float = 0
    cost: float = 0
    stock: int = 0
    min_stock: int = 5
    unit: Optional[str] = "pcs"
    image: Optional[str] = ""
    description: Optional[str] = ""
    active: bool = True


class ReorderInput(BaseModel):
    ids: List[str]


class SaleItem(BaseModel):
    product_id: str
    name: str
    price: float
    qty: int
    cost: float = 0
    note: Optional[str] = ""


class SaleInput(BaseModel):
    items: List[SaleItem]
    discount: float = 0
    tax_rate: float = 0
    payment_method: PAYMENT_METHODS
    paid_amount: float = 0
    customer_name: Optional[str] = ""
    customer_id: Optional[str] = None
    order_id: Optional[str] = None
    channel: Optional[str] = "Toko"


class StockInput(BaseModel):
    product_id: str
    type: Literal["Masuk", "Keluar", "Penyesuaian", "Opname"]
    qty: int
    note: Optional[str] = ""


class SettingsInput(BaseModel):
    business_name: Optional[str] = None
    address: Optional[str] = None
    phone: Optional[str] = None
    currency: Optional[str] = None
    tax_rate: Optional[float] = None
    receipt_footer: Optional[str] = None
    logo: Optional[str] = None
    print_mode: Optional[str] = None
    paper_width: Optional[str] = None
    printers: Optional[list] = None
    active_printer: Optional[str] = None


class CustomerInput(BaseModel):
    name: str
    phone: Optional[str] = ""
    email: Optional[str] = ""
    address: Optional[str] = ""


class SupplierInput(BaseModel):
    name: str
    phone: Optional[str] = ""
    email: Optional[str] = ""
    address: Optional[str] = ""


class POItem(BaseModel):
    product_id: str
    name: str
    qty: int
    cost: float = 0


class PurchaseOrderInput(BaseModel):
    supplier_id: Optional[str] = None
    supplier_name: Optional[str] = ""
    items: List[POItem]
    note: Optional[str] = ""


class SupplierRef(BaseModel):
    supplier_id: Optional[str] = None


class HeldOrderInput(BaseModel):
    label: str
    items: List[SaleItem]
    discount: float = 0


class CustomOrderInput(BaseModel):
    customer_id: Optional[str] = None
    customer_name: Optional[str] = ""
    items: List[SaleItem]
    discount: float = 0
    tax_rate: float = 0
    deposit_amount: float = 0
    deposit_method: PAYMENT_METHODS = "Tunai"
    order_type: str = "Reguler"
    note: Optional[str] = ""
    channel: Optional[str] = "Toko"


class OrderDepositInput(BaseModel):
    deposit_amount: float
    deposit_method: PAYMENT_METHODS = "Tunai"


class UpdateOrderInput(BaseModel):
    items: List[SaleItem]
    discount: float = 0
    tax_rate: float = 0
    customer_name: Optional[str] = ""
    order_type: str = "Reguler"


class SettleOrderInput(BaseModel):
    payment_method: PAYMENT_METHODS
    paid_amount: float = 0


class FinanceCategoryInput(BaseModel):
    name: str
    type: Literal["expense", "income"]


class FinanceEntryInput(BaseModel):
    category: str
    amount: float
    note: Optional[str] = ""
    date: Optional[str] = None


class GalleryInput(BaseModel):
    src: str
    label: str
    tag: Optional[str] = ""
    span: Optional[str] = ""
    sort_order: Optional[int] = 0

File: web/lib/schemas.ts
// Zod schemas mirroring backend/app/schemas.py
import { z } from 'zod';

export const PAYMENT_METHODS = ['Tunai', 'BCA TOKO', 'BRI TOKO', 'BCA ADMIN (ELIS)', 'QRIS', 'E-Wallet'] as const;
export const ROLES = ['Owner', 'Manager', 'Kasir', 'Gudang'] as const;

export const loginSchema = z.object({ username: z.string(), password: z.string() });
export const changePasswordSchema = z.object({ current_password: z.string(), new_password: z.string() });

export const userCreateSchema = z.object({
  username: z.string(),
  password: z.string(),
  name: z.string(),
  role: z.enum(ROLES),
});
export const userUpdateSchema = z.object({
  name: z.string().optional(),
  role: z.enum(ROLES).optional(),
  password: z.string().optional(),
  active: z.boolean().optional(),
});

export const categoryInputSchema = z.object({
  name: z.string(),
  color: z.string().optional().default('#2563EB'),
  image: z.string().optional().default(''),
});

export const productInputSchema = z.object({
  name: z.string(),
  sku: z.string().optional().default(''),
  barcode: z.string().optional().default(''),
  category_id: z.string().nullable().optional(),
  price: z.number().default(0),
  cost: z.number().default(0),
  stock: z.number().int().default(0),
  min_stock: z.number().int().default(5),
  unit: z.string().optional().default('pcs'),
  image: z.string().optional().default(''),
  description: z.string().optional().default(''),
  active: z.boolean().default(true),
});

export const reorderInputSchema = z.object({ ids: z.array(z.string()) });

export const saleItemSchema = z.object({
  product_id: z.string(),
  name: z.string(),
  price: z.number(),
  qty: z.number().int(),
  cost: z.number().default(0),
  note: z.string().optional().default(''),
});

export const saleInputSchema = z.object({
  items: z.array(saleItemSchema),
  discount: z.number().default(0),
  tax_rate: z.number().default(0),
  payment_method: z.enum(PAYMENT_METHODS),
  paid_amount: z.number().default(0),
  customer_name: z.string().optional().default(''),
  customer_id: z.string().nullable().optional(),
  order_id: z.string().nullable().optional(),
  channel: z.string().optional().default('Toko'),
});

export const stockInputSchema = z.object({
  product_id: z.string(),
  type: z.enum(['Masuk', 'Keluar', 'Penyesuaian', 'Opname']),
  qty: z.number().int(),
  note: z.string().optional().default(''),
});

export const settingsInputSchema = z.object({
  business_name: z.string().nullable().optional(),
  address: z.string().nullable().optional(),
  phone: z.string().nullable().optional(),
  currency: z.string().nullable().optional(),
  tax_rate: z.number().nullable().optional(),
  receipt_footer: z.string().nullable().optional(),
  logo: z.string().nullable().optional(),
  print_mode: z.string().nullable().optional(),
  paper_width: z.string().nullable().optional(),
  printers: z.any().nullable().optional(),
  active_printer: z.string().nullable().optional(),
}).passthrough();

export const customerInputSchema = z.object({
  name: z.string(),
  phone: z.string().optional().default(''),
  email: z.string().optional().default(''),
  address: z.string().optional().default(''),
});

export const supplierInputSchema = customerInputSchema;

export const poItemSchema = z.object({
  product_id: z.string(),
  name: z.string(),
  qty: z.number().int(),
  cost: z.number().default(0),
});

export const purchaseOrderInputSchema = z.object({
  supplier_id: z.string().nullable().optional(),
  supplier_name: z.string().optional().default(''),
  items: z.array(poItemSchema),
  note: z.string().optional().default(''),
});

export const supplierRefSchema = z.object({ supplier_id: z.string().nullable().optional() });

export const heldOrderInputSchema = z.object({
  label: z.string(),
  items: z.array(saleItemSchema),
  discount: z.number().default(0),
});

export const customOrderInputSchema = z.object({
  customer_id: z.string().nullable().optional(),
  customer_name: z.string().optional().default(''),
  items: z.array(saleItemSchema),
  discount: z.number().default(0),
  tax_rate: z.number().default(0),
  deposit_amount: z.number().default(0),
  deposit_method: z.enum(PAYMENT_METHODS).default('Tunai'),
  order_type: z.string().default('Reguler'),
  note: z.string().optional().default(''),
  channel: z.string().optional().default('Toko'),
});

export const orderDepositSchema = z.object({
  deposit_amount: z.number(),
  deposit_method: z.enum(PAYMENT_METHODS).default('Tunai'),
});

export const updateOrderSchema = z.object({
  items: z.array(saleItemSchema),
  discount: z.number().default(0),
  tax_rate: z.number().default(0),
  customer_name: z.string().optional().default(''),
  order_type: z.string().default('Reguler'),
});

export const settleOrderSchema = z.object({
  payment_method: z.enum(PAYMENT_METHODS),
  paid_amount: z.number().default(0),
});

export const financeCategoryInputSchema = z.object({
  name: z.string(),
  type: z.enum(['expense', 'income']),
});

export const financeEntryInputSchema = z.object({
  category: z.string(),
  amount: z.number(),
  note: z.string().optional().default(''),
  date: z.string().nullable().optional(),
});

export const galleryInputSchema = z.object({
  src: z.string(),
  label: z.string(),
  tag: z.string().optional().default(''),
  span: z.string().optional().default(''),
  sort_order: z.number().int().nullable().optional().default(0),
});

export function parseBody<T extends z.ZodTypeAny>(schema: T, body: unknown): z.infer<T> {
  const r = schema.safeParse(body);
  if (!r.success) {
    const first = r.error.issues[0];
    const path = first.path.join('.');
    throw new Error(`${path}: ${first.message}`);
  }
  return r.data;
}
